src/shared/machine-id-setup.c

```diff
@@ -27,22 +27,17 @@
 #include "virt.h"
 
 static int generate_machine_id(const char *root, sd_id128_t *ret) {
-        const char *dbus_machine_id;
-        _cleanup_close_ int fd = -EBADF;
+        _cleanup_close_ int dir_fd = -EBADF;
         int r;
 
         assert(ret);
 
         /* First, try reading the D-Bus machine id, unless it is a symlink */
-        dbus_machine_id = prefix_roota(root, "/var/lib/dbus/machine-id");
-        fd = open(dbus_machine_id, O_RDONLY|O_CLOEXEC|O_NOCTTY|O_NOFOLLOW);
-        if (fd >= 0) {
-                if (id128_read_fd(fd, ID128_FORMAT_PLAIN, ret) >= 0) {
-                        log_info("Initializing machine ID from D-Bus machine ID.");
-                        return 0;
-                }
-
-                fd = safe_close(fd);
+        dir_fd = open(empty_to_root(root), O_CLOEXEC|O_DIRECTORY|O_PATH);
```

Comment:
We should `chase_and_open()` "var/lib/dbus/machine-id" here and then just id128_read() the fd. WIth the current approach, if var/, lib/ or dbus/ are symlinks, we'll end up escaping the root directory since `ID128_NOFOLLOW` only applies to the final component of the path.